types/mempool/mempool_test.go

```diff
@@ -0,0 +1,659 @@
+	}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+
+	// removing a tx not in the mempool should error
+	mp = mempool.NewPriorityMempool()
+	require.NoError(t, mp.Insert(ctx, txs[0]))
+	require.ErrorIs(t, mp.Remove(txs[1]), mempool.ErrTxNotFound)
+	mempool.DebugPrintKeys(mp)
+
+	// inserting a tx with a different priority should overwrite the old tx
+	newPriorityTx := testTx{
+		address:  txs[0].address,
+		priority: txs[0].priority + 1,
+		nonce:    txs[0].nonce,
+	}
+	require.NoError(t, mp.Insert(ctx, newPriorityTx))
+	require.Equal(t, 1, mp.CountTx())
+}
+
+type txSpec struct {
+	i int
+	h int
+	p int
+	n int
+	a sdk.AccAddress
+}
+
+func (tx txSpec) String() string {
+	return fmt.Sprintf("[tx i: %d, a: %s, p: %d, n: %d]", tx.i, tx.a, tx.p, tx.n)
+}
+
+func TestOutOfOrder(t *testing.T) {
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 2)
+	sa := accounts[0].Address
+	sb := accounts[1].Address
+
+	outOfOrders := [][]testTx{
+		{
+			{priority: 20, nonce: 1, address: sa},
+			{priority: 21, nonce: 4, address: sa},
+			{priority: 15, nonce: 1, address: sb},
+			{priority: 8, nonce: 3, address: sa},
+			{priority: 6, nonce: 2, address: sa},
+		},
+		{
+			{priority: 15, nonce: 1, address: sb},
+			{priority: 20, nonce: 1, address: sa},
+			{priority: 21, nonce: 4, address: sa},
+			{priority: 8, nonce: 3, address: sa},
+			{priority: 6, nonce: 2, address: sa},
+		}}
+
+	for _, outOfOrder := range outOfOrders {
+		var mtxs []mempool.Tx
+		for _, mtx := range outOfOrder {
+			mtxs = append(mtxs, mtx)
+		}
+		err := validateOrder(mtxs)
+		require.Error(t, err)
+	}
+
+	seed := time.Now().UnixNano()
+	randomTxs := genRandomTxs(seed, 1000, 10)
+	var rmtxs []mempool.Tx
+	for _, rtx := range randomTxs {
+		rmtxs = append(rmtxs, rtx)
+	}
+
+	require.Error(t, validateOrder(rmtxs))
+
+}
+
+func (s *MempoolTestSuite) TestTxOrder() {
+	t := s.T()
+	ctx := sdk.NewContext(nil, tmproto.Header{}, false, log.NewNopLogger())
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 5)
+	sa := accounts[0].Address
+	sb := accounts[1].Address
+	sc := accounts[2].Address
+
+	tests := []struct {
+		txs   []txSpec
+		order []int
+		fail  bool
+	}{
+		{
+			txs: []txSpec{
+				{p: 21, n: 4, a: sa},
+				{p: 8, n: 3, a: sa},
+				{p: 6, n: 2, a: sa},
+				{p: 15, n: 1, a: sb},
+				{p: 20, n: 1, a: sa},
+			},
+			order: []int{4, 3, 2, 1, 0},
+		},
+		{
+			txs: []txSpec{
+				{p: 3, n: 0, a: sa},
+				{p: 5, n: 1, a: sa},
+				{p: 9, n: 2, a: sa},
+				{p: 6, n: 0, a: sb},
+				{p: 5, n: 1, a: sb},
+				{p: 8, n: 2, a: sb},
+			},
+			order: []int{3, 4, 5, 0, 1, 2},
+		},
+		{
+			txs: []txSpec{
+				{p: 21, n: 4, a: sa},
+				{p: 15, n: 1, a: sb},
+				{p: 20, n: 1, a: sa},
+			},
+			order: []int{2, 0, 1}},
+		{
+			txs: []txSpec{
+				{p: 50, n: 3, a: sa},
+				{p: 30, n: 2, a: sa},
+				{p: 10, n: 1, a: sa},
+				{p: 15, n: 1, a: sb},
+				{p: 21, n: 2, a: sb},
+			},
+			order: []int{3, 4, 2, 1, 0},
+		},
```

Comment:
how about this case:
```go
		{
			txs: []txSpec{
				{p: 5, n: 1, a: sa},
				{p: 10, n: 2, a: sa},
				{p: 5, n: 1, a: sb},
				{p: 99, n: 2, a: sb},
			},
			order: []int{2, 3, 0, 1},
		},
```

expected: I would say 2,3,0,1, coz we'd want to get the 99 priority in asap. Or do we not support that?
actual: 0,1,2,3

Reply:
This is great, thank you :pray:  I think we need some weighting to resolve priority ties between senders.

Reply:
That's a great case @amaurym. Since you can make a split decision between TX0 and TX2, it's preferable to pick TX2 as that'll allow TX3 to be picked next as that's higher priority than if you went with TX0 first.

> I think we need some weighting to resolve priority ties between senders.

Yes.

Reply:
Here's a sister test case:

```go
		{
			txs: []txSpec{
				{p: 6, n: 1, a: sa},
				{p: 10, n: 2, a: sa},
				{p: 5, n: 1, a: sb},
				{p: 99, n: 2, a: sb},
			},
		},
```

There's no priority tie for the first tx in each sender bucket, but I would argue the ordering should still be 2,3,0,1. Although 0,1,2,3 could also be justifiable. In any case, would love to see these 2 test cases as to understand the spec of this mempool impl.

Reply:
A similar case was studied by @JeancarloBarrios and I.  In this case we'd say that 0,1,2,3 is the correct order, reasoning that p=6 should *always* run before p=5 no matter what priority is behind it nonce-wise.

I'll add this one too though.